File: internal/prepare/prepare.go
package prepare

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type ScanInput struct {
	BaseDir string
	Images  []string
	Videos  []string
}

type Options struct {
	RenameImages bool
	RenameVideos bool
	Overwrite    bool
}

type Operation struct {
	Source      string `json:"source"`
	Destination string `json:"destination"`
	Kind        string `json:"kind"`
}

type Result struct {
	RenamedImages int    `json:"renamedImages"`
	RenamedVideos int    `json:"renamedVideos"`
	TotalFiles    int    `json:"totalFiles"`
	OutputDir     string `json:"outputDir"`
}

type ProgressFunc func(current int, total int, message string)

func PlanOperations(scan ScanInput, outputDir string, options Options) ([]Operation, error) {
	base := filepath.Clean(scan.BaseDir)
	var ops []Operation

	imageOps, err := planKind(base, outputDir, scan.Images, "image", options.RenameImages, options.Overwrite)
	if err != nil {
		return nil, err
	}
	videoOps, err := planKind(base, outputDir, scan.Videos, "video", options.RenameVideos, options.Overwrite)
	if err != nil {
		return nil, err
	}

	ops = append(ops, imageOps...)
	ops = append(ops, videoOps...)
	return ops, nil
}

func ExecuteOperations(ops []Operation, overwrite bool, progress ProgressFunc) (Result, error) {
	result := Result{TotalFiles: len(ops)}
	if len(ops) > 0 {
		result.OutputDir = filepath.Dir(ops[0].Destination)
	}

	for index, op := range ops {
		if progress != nil {
			progress(index+1, len(ops), fmt.Sprintf("%s -> %s", filepath.Base(op.Source), filepath.Base(op.Destination)))
		}
		if !overwrite {
			if _, err := os.Stat(op.Destination); err == nil {
				return result, fmt.Errorf("destination already exists: %s", op.Destination)
			} else if !errors.Is(err, os.ErrNotExist) {
				return result, err
			}
		}
		if err := copyFile(op.Source, op.Destination); err != nil {
			return result, err
		}
		switch op.Kind {
		case "image":
			result.RenamedImages++
		case "video":
			result.RenamedVideos++
		}
	}
	return result, nil
}

func copyFile(source string, destination string) error {
	input, err := os.Open(source)
	if err != nil {
		return err
	}
	defer input.Close()

	if err := os.MkdirAll(filepath.Dir(destination), 0o755); err != nil {
		return err
	}
	output, err := os.Create(destination)
	if err != nil {
		return err
	}
	defer output.Close()

	if _, err := io.Copy(output, input); err != nil {
		return err
	}
	return output.Close()
}

func planKind(base, outputDir string, files []string, kind string, rename bool, overwrite bool) ([]Operation, error) {
	groups := map[string][]string{}
	for _, file := range files {
		groups[filepath.Dir(file)] = append(groups[filepath.Dir(file)], file)
	}

	parents := make([]string, 0, len(groups))
	for parent := range groups {
		parents = append(parents, parent)
		sort.Strings(groups[parent])
	}
	sort.Strings(parents)

	var ops []Operation
	for _, parent := range parents {
		rel := "."
		if filepath.Clean(parent) != base {
			nextRel, err := filepath.Rel(base, parent)
			if err != nil {
				return nil, err
			}
			rel = nextRel
		}
		for i, file := range groups[parent] {
			name := filepath.Base(file)
			if rename {
				ext := strings.ToLower(filepath.Ext(file))
				if kind == "video" {
					name = fmt.Sprintf("video%03d%s", i+1, ext)
				} else {
					name = fmt.Sprintf("%04d%s", i+1, ext)
				}
			}
			destDir := parent
			if !overwrite {
				destDir = outputDir
				if rel != "." {
					destDir = filepath.Join(outputDir, rel)
				}
			}
			ops = append(ops, Operation{
				Source:      file,
				Destination: filepath.Join(destDir, name),
				Kind:        kind,
			})
		}
	}
	return ops, nil
}

File: src/core/prepare.rs
use std::fs::{self, File};
use std::io;
use std::path::Path;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ScanInput {
    #[serde(rename = "BaseDir")]
    pub base_dir: String,
    #[serde(rename = "Images")]
    pub images: Vec<String>,
    #[serde(rename = "Videos")]
    pub videos: Vec<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Options {
    #[serde(rename = "RenameImages")]
    pub rename_images: bool,
    #[serde(rename = "RenameVideos")]
    pub rename_videos: bool,
    #[serde(rename = "Overwrite")]
    pub overwrite: bool,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Operation {
    pub source: String,
    pub destination: String,
    pub kind: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, Default)]
pub struct PrepareResult {
    #[serde(rename = "renamedImages")]
    pub renamed_images: i32,
    #[serde(rename = "renamedVideos")]
    pub renamed_videos: i32,
    #[serde(rename = "totalFiles")]
    pub total_files: i32,
    #[serde(rename = "outputDir")]
    pub output_dir: String,
}

pub type ProgressFunc = Option<Box<dyn Fn(usize, usize, String) + Send + Sync>>;

pub fn plan_operations(
    scan: ScanInput,
    output_dir: &str,
    opts: Options,
) -> anyhow::Result<Vec<Operation>> {
    let base = Path::new(&scan.base_dir)
        .to_string_lossy()
        .replace('\\', "/");
    let mut ops = Vec::new();
    ops.extend(plan_kind(
        &base,
        output_dir,
        &scan.images,
        "image",
        opts.rename_images,
        opts.overwrite,
    )?);
    ops.extend(plan_kind(
        &base,
        output_dir,
        &scan.videos,
        "video",
        opts.rename_videos,
        opts.overwrite,
    )?);
    Ok(ops)
}

fn plan_kind(
    base: &str,
    output_dir: &str,
    files: &[String],
    kind: &str,
    rename: bool,
    overwrite: bool,
) -> anyhow::Result<Vec<Operation>> {
    use std::collections::BTreeMap;
    let mut groups: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for f in files {
        let p = Path::new(f);
        let dir = p
            .parent()
            .map(|d| d.to_string_lossy().replace('\\', "/"))
            .unwrap_or_else(|| ".".to_string());
        groups.entry(dir).or_default().push(f.clone());
    }
    let mut parents: Vec<_> = groups.keys().cloned().collect();
    parents.sort();
    let mut ops = Vec::new();
    for parent in parents {
        let rel = if parent == base {
            ".".to_string()
        } else {
            // compute rel
            let base_p = Path::new(base);
            let par_p = Path::new(&parent);
            match par_p.strip_prefix(base_p) {
                Ok(r) => r.to_string_lossy().replace('\\', "/"),
                _ => parent.clone(),
            }
        };
        let mut fs = groups.get(&parent).unwrap().clone();
        fs.sort();
        for (i, file) in fs.into_iter().enumerate() {
            let name = if rename {
                let ext = Path::new(&file)
                    .extension()
                    .and_then(|e| e.to_str())
                    .unwrap_or("");
                let ext = format!(".{}", ext.to_ascii_lowercase());
                if kind == "video" {
                    format!("video{:03}{}", i + 1, ext)
                } else {
                    format!("{:04}{}", i + 1, ext)
                }
            } else {
                Path::new(&file)
                    .file_name()
                    .unwrap()
                    .to_string_lossy()
                    .to_string()
            };
            let dest_dir = if overwrite {
                parent.clone()
            } else if rel == "." || rel.is_empty() {
                output_dir.to_string()
            } else {
                format!(
                    "{}/{}",
                    output_dir.trim_end_matches('/'),
                    rel.trim_start_matches('/')
                )
            };
            let dest = format!("{}/{}", dest_dir.trim_end_matches('/'), name);
            ops.push(Operation {
                source: file,
                destination: dest.replace('\\', "/"),
                kind: kind.to_string(),
            });
        }
    }
    Ok(ops)
}

pub fn execute_operations(
    ops: Vec<Operation>,
    overwrite: bool,
    progress: ProgressFunc,
) -> anyhow::Result<PrepareResult> {
    let mut res = PrepareResult {
        total_files: ops.len() as i32,
        ..Default::default()
    };
    if let Some(first) = ops.first() {
        if let Some(d) = Path::new(&first.destination).parent() {
            res.output_dir = d.to_string_lossy().to_string();
        }
    }
    for (idx, op) in ops.into_iter().enumerate() {
        if let Some(pf) = &progress {
            pf(
                idx + 1,
                res.total_files as usize,
                format!(
                    "{} -> {}",
                    Path::new(&op.source).file_name().unwrap().to_string_lossy(),
                    Path::new(&op.destination)
                        .file_name()
                        .unwrap()
                        .to_string_lossy()
                ),
            );
        }
        if !overwrite {
            if Path::new(&op.destination).exists() {
                return Err(anyhow::anyhow!(
                    "destination already exists: {}",
                    op.destination
                ));
            }
        }
        if let Some(parent) = Path::new(&op.destination).parent() {
            fs::create_dir_all(parent)?;
        }
        fs::copy(&op.source, &op.destination)?;
        match op.kind.as_str() {
            "image" => res.renamed_images += 1,
            "video" => res.renamed_videos += 1,
            _ => {}
        }
    }
    Ok(res)
}
